### PR TITLE
Add AVX512 support to k-nn FAISS

### LINKED ISSUE #2056: [FEATURE] Add support for FAISS AVX512
**Is your feature request related to a problem?**
Improve performance of vector search and indexing when using the FAISS library, by using AVX512 accelerators which are supported in faiss. 

**What solution would you like?**
Add support to use AVX512, initially as an optional feature and eventually enable by default on machines where AVX512 is supported on subsequent OpenSearch releases.
This will be a 2 part change: 
1) Adds necessary code in k-NN to check for AVX512 support on the machine.
2) A separate change made in FAISS to add missing functions support for AVX512.

### PR DESCRIPTION
**Pull request by akashsha1**
### Description
This change adds support to speed up vector search and indexing in faiss using AVX512 hardware accelerator.

### Related Issues
Resolves #2056

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -19,6 +19,7 @@ buildscript {
         isSnapshot = "true" == System.getProperty("build.snapshot", "true")
         simd_enabled = System.getProperty("simd.enabled", "true")
```
Comment (naveentatikonda) [inline build.gradle:20]: Can you change `simd.enabled` to `avx2.enabled` ?

File: build.gradle [hunk 2]
```diff
@@ -317,6 +318,7 @@ task cmakeJniLib(type:Exec) {
     args.add(".")
     args.add("-DKNN_PLUGIN_VERSION=${opensearch_version}")
     args.add("-DSIMD_ENABLED=${simd_enabled}")
```
Comment (naveentatikonda) [inline build.gradle:320]: Same here change simd to avx2

File: jni/cmake/init-faiss.cmake [hunk 3]
```diff
@@ -85,9 +85,17 @@ if(NOT DEFINED SIMD_ENABLED)
     set(SIMD_ENABLED true)   # set default value as true if the argument is not set
```
Comment (naveentatikonda) [inline jni/cmake/init-faiss.cmake:85]: same as above comments

File: jni/cmake/init-faiss.cmake [hunk 4]
```diff
@@ -85,9 +85,17 @@ if(NOT DEFINED SIMD_ENABLED)
...
 endif()
 
-if(${CMAKE_SYSTEM_NAME} STREQUAL Windows OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "aarch64" OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm64" OR NOT ${SIMD_ENABLED})
+if(NOT DEFINED AVX512_ENABLED)
+    set(AVX512_ENABLED true)   # set default value as true if the argument is not set
+endif()
+
+if(${CMAKE_SYSTEM_NAME} STREQUAL Windows OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "aarch64" OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm64" OR NOT SIMD_ENABLED)
```
Comment (naveentatikonda) [inline jni/cmake/init-faiss.cmake:92]:
```suggestion
if(${CMAKE_SYSTEM_NAME} STREQUAL Windows OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "aarch64" OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm64" OR NOT AVX2_ENABLED OR NOT AVX512_ENABLED)
```

File: jni/cmake/init-faiss.cmake [hunk 5]
```diff
@@ -85,9 +85,17 @@ if(NOT DEFINED SIMD_ENABLED)
...
     set(FAISS_OPT_LEVEL generic)    # Keep optimization level as generic on Windows OS as it is not supported due to MINGW64 compiler issue. Also, on aarch64 avx2 is not supported.
     set(TARGET_LINK_FAISS_LIB faiss)
+elseif(${CMAKE_SYSTEM_NAME} STREQUAL Linux AND SIMD_ENABLED AND AVX512_ENABLED)
```
Comment (naveentatikonda) [inline jni/cmake/init-faiss.cmake:95]:
```suggestion
elseif(${CMAKE_SYSTEM_NAME} STREQUAL Linux AND AVX512_ENABLED)
```

File: jni/cmake/init-faiss.cmake [hunk 6]
```diff
@@ -85,9 +85,17 @@ if(NOT DEFINED SIMD_ENABLED)
...
+    set(FAISS_OPT_LEVEL avx512)       # Keep optimization level as avx512 to improve performance on Linux. This is not present on mac systems.
```
Comment (naveentatikonda) [inline jni/cmake/init-faiss.cmake:96]: Can you pls mention in the comment that it is not supported for Windows OS

File: src/main/java/org/opensearch/knn/jni/FaissService.java [hunk 7]
```diff
@@ -35,10 +36,16 @@ class FaissService {
     static {
         AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
 
-            // Even if the underlying system supports AVX2, users can override and disable it by using the
+            // Even if the underlying system supports AVX512 and AVX2, users can override and disable it by using the
             // 'knn.faiss.avx2.disabled' setting by setting it to true in the opensearch.yml configuration
```
Comment (naveentatikonda) [inline src/main/java/org/opensearch/knn/jni/FaissService.java:40]: Also mention about `knn.faiss.avx512.disabled` setting in the comments

File: src/main/java/org/opensearch/knn/jni/FaissService.java [hunk 8]
```diff
@@ -35,10 +36,16 @@ class FaissService {
...
-            if (!isFaissAVX2Disabled() && isAVX2SupportedBySystem()) {
-                System.loadLibrary(KNNConstants.FAISS_AVX2_JNI_LIBRARY_NAME);
+            if (!isFaissAVX2Disabled()) {
```
Comment (naveentatikonda) [inline src/main/java/org/opensearch/knn/jni/FaissService.java:41]:
```suggestion
            if (!isFaissAVX5122Disabled() && isAVX512SupportedBySystem()) {
                System.loadLibrary(KNNConstants.FAISS_AVX512_JNI_LIBRARY_NAME);
                } else if (!isFaissAVX2Disabled() && isAVX2SupportedBySystem()) {
                System.loadLibrary(KNNConstants.FAISS_AVX2_JNI_LIBRARY_NAME);
                } else {
                    System.loadLibrary(KNNConstants.FAISS_JNI_LIBRARY_NAME);
                }
```

File: src/main/java/org/opensearch/knn/jni/PlatformUtils.java [hunk 9]
```diff
@@ -80,4 +81,38 @@ public static boolean isAVX2SupportedBySystem() {
         }
         return false;
     }
+
+    public static boolean isAVX512SupportedBySystem() {
+
+        if (!Platform.isIntel() || Platform.isMac()) {
```
Comment (naveentatikonda) [inline src/main/java/org/opensearch/knn/jni/PlatformUtils.java:87]: return false for Windows platform too as we are not supporting it

**Comment by naveentatikonda** [comment]
Can you also please update build script to build all 3 versions of faiss lib for the distribution. Pls change [this block of code](https://github.com/opensearch-project/k-NN/blob/main/scripts/build.sh#L123-L128) like shown below 
```
if [ "$PLATFORM" != "windows" ] && [ "$ARCHITECTURE" = "x64" ]; then
  echo "Building k-NN library after enabling AVX2"
  # Skip applying patches as patches were applied already from previous :buildJniLib task
  # If we apply patches again, it fails with conflict
  ./gradlew :buildJniLib -Davx2.enabled=true -Dbuild.lib.commit_patches=false -Dbuild.lib.apply_patches=false

  echo "Building k-NN library after enabling AVX512"
  ./gradlew :buildJniLib -Davx512.enabled=true -Dbuild.lib.commit_patches=false -Dbuild.lib.apply_patches=false
fi
```

**Comment by naveentatikonda** [comment]
@akashsha1 You also need to add the avx512 faiss library in [plugin-security.policy](https://github.com/opensearch-project/k-NN/blob/main/src/main/plugin-metadata/plugin-security.policy#L5) to provide runtime permissions for loading the library. Pls test end to end after making all these changes. Thanks!

File: .github/workflows/CI.yml [hunk 10]
```diff
@@ -73,13 +73,17 @@ jobs:
         # switching the user, as OpenSearch cluster can only be started as root/Administrator on linux-deb/linux-rpm/windows-zip.
         run: |
           chown -R 1000:1000 `pwd`
-          if lscpu  | grep -i avx2
+          if lscpu  | grep -i avx512f | grep -i avx512cd | grep -i avx512vl | grep -i avx512dq | grep -i avx512bw
```
Comment (naveentatikonda) [inline .github/workflows/CI.yml:76]:
```suggestion
          if lscpu  | grep -i avx512          
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -16,6 +16,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased 2.x](https://github.com/opensearch-project/k-NN/compare/2.17...2.x)
 ### Features
+* Add AVX512 support to k-NN for FAISS library
```
Comment (naveentatikonda) [inline CHANGELOG.md:19]:
```suggestion
* Add AVX512 support to k-NN for FAISS library [#2069](https://github.com/opensearch-project/k-NN/pull/2069)
```

File: .github/workflows/CI.yml (hunk 10 above)
Comment (akashsha1) [reply .github/workflows/CI.yml:76]: that won't be accurate. We want all 5 flags to be present. The above check will return success, even if only 3/5 avx512 flags are found.
AVX512 has been incrementally improved over the years, so older systems or client machines may not have all of them.

**Comment by naveentatikonda** [comment]
@akashsha1 Looks like some of the commits were not properly signed causing the DCO to fail. Do you want to rebase and squash everything into a single commit and sign that commit ?

image

File: scripts/build.sh [hunk 12]
```diff
@@ -118,13 +118,16 @@ fi
...
 ./gradlew build --no-daemon --refresh-dependencies -x integTest -x test -Dopensearch.version=$VERSION -Dbuild.snapshot=$SNAPSHOT -Dbuild.version_qualifier=$QUALIFIER -Dbuild.lib.commit_patches=false
-./gradlew :buildJniLib -Dsimd.enabled=false -Dbuild.lib.commit_patches=false
+./gradlew :buildJniLib -Davx512.enabled=false -Davx2.enabled=false -Dbuild.lib.commit_patches=false
 
 if [ "$PLATFORM" != "windows" ] && [ "$ARCHITECTURE" = "x64" ]; then
   echo "Building k-NN library after enabling AVX2"
   # Skip applying patches as patches were applied already from previous :buildJniLib task
   # If we apply patches again, it fails with conflict
-  ./gradlew :buildJniLib -Dsimd.enabled=true -Dbuild.lib.commit_patches=false -Dbuild.lib.apply_patches=false
+  ./gradlew :buildJniLib -Davx2.enabled=true -Dbuild.lib.commit_patches=false -Dbuild.lib.apply_patches=false
```
Comment (naveentatikonda) [inline scripts/build.sh:127]:
```suggestion
  ./gradlew :buildJniLib -Davx2.enabled=true -Davx512.enabled=false -Dbuild.lib.commit_patches=false -Dbuild.lib.apply_patches=false
```

Comment (naveentatikonda) [reply scripts/build.sh:127]: We need to disable avx512 and enable avx2 to build avx2 version of faiss lib

**Comment by naveentatikonda** [comment]
@akashsha1 We need to add the CPU flag detection logic in these 2 workflows and pass the gradle flags to fix the failing checks
https://github.com/opensearch-project/k-NN/blob/main/.github/workflows/backwards_compatibility_tests_workflow.yml
https://github.com/opensearch-project/k-NN/blob/main/.github/workflows/test_security.yml

**Comment by akashsha1** [comment]
Hi @naveentatikonda - please consider this PR: https://github.com/opensearch-project/k-NN/pull/2110 in favor of the current one. The rebasing due to DCO sign-off got messy, so I created a new PR instead. It has all feedback addressed to this point, except the workflow failures which needs to be resolved.

**Comment by akashsha1** [comment]
> @akashsha1 We need to add the CPU flag detection logic in these 2 workflows and pass the gradle flags to fix the failing checks https://github.com/opensearch-project/k-NN/blob/main/.github/workflows/backwards_compatibility_tests_workflow.yml https://github.com/opensearch-project/k-NN/blob/main/.github/workflows/test_security.yml

This needs to be addressed in https://github.com/opensearch-project/k-NN/pull/2110 ..

**Pull request closed without merging**